core/services/workflows/metering/metering.go

```diff
@@ -178,141 +180,84 @@ func (r *Report) Reserve(ctx context.Context) error {
 	return nil
 }
 
-// ConvertToBalance converts a resource dimensions amount to a credit amount.
-func (r *Report) ConvertToBalance(fromUnit string, amount decimal.Decimal) (decimal.Decimal, error) {
-	if !r.ready {
-		return decimal.Zero, ErrNoReserve
-	}
-
-	bal, err := r.balance.ConvertToBalance(fromUnit, amount)
-	if err != nil {
-		// Fail open, continue optimistically
-		r.switchToMeteringMode(fmt.Errorf("failed to convert to balance [%s]: %w", fromUnit, err))
-	}
-
-	return bal, nil
-}
-
-// Deduct earmarks an amount of local universal credit balance. We expect to only set this value once - an error is
-// returned if a step would be overwritten.
-func (r *Report) Deduct(ref string, amount decimal.Decimal) error {
-	r.mu.Lock()
-	defer r.mu.Unlock()
+type DeductOpt func(string, *Report) ([]capabilities.SpendLimit, error)
 
-	if !r.ready {
-		return ErrNoReserve
-	}
+func WithNativeValue(
+	spendType string,
+	amount decimal.Decimal,
+) func(string, *Report) ([]capabilities.SpendLimit, error) {
+	return func(ref string, r *Report) ([]capabilities.SpendLimit, error) {
+		bal, err := r.balance.ConvertToBalance(spendType, amount)
+		if err != nil {
+			// Fail open, continue optimistically
+			r.switchToMeteringMode(fmt.Errorf("failed to convert to balance [%s]: %w", spendType, err))
+		}
 
-	if _, ok := r.steps[ref]; ok {
-		return ErrStepDeductExists
-	}
+		r.steps[ref] = ReportStep{
+			Deduction: bal,
+			Spends:    nil,
+		}
 
-	r.steps[ref] = ReportStep{
-		Deduction: amount,
-		Spends:    nil,
-	}
+		// if in metering mode, exit early without modifying local balance
+		if r.meteringMode {
+			return []capabilities.SpendLimit{}, nil
+		}
 
-	// if in metering mode, exit early without modifying local balance
-	if r.meteringMode {
-		return nil
+		return []capabilities.SpendLimit{}, r.balance.Minus(bal)
 	}
-
-	return r.balance.Minus(amount)
 }
 
-// CreditToSpendingLimits returns a slice of spend limits where the amount is applied to the spend types from the
-// provided info. Amount should be specified in universal credits and will be converted to spend type credits within
-// this function.
-func (r *Report) CreditToSpendingLimits(
+func WithDerivedValue(
```

Comment:
Not a strongly held comment:

What about naming this something like ByDerivedAvailability - AKA Deduct.ByDerivedAvailability

Gives some hint in the name on what it is derived using